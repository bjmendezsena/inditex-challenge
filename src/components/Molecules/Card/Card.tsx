import { useMultiStyleConfig as useChakraMultiStyleConfig } from "@chakra-ui/system";
import { VStack as ChakraVStack } from "@chakra-ui/react";
import {
  Card as ChakraCard,
  CardProps as ChakraCardProps,
  CardBody as ChakraCardBody,
  CardBodyProps as ChakraCardBodyProps,
  CardHeader as ChakraCardHeader,
  CardHeaderProps as ChakraCardHeaderProps,
  CardFooter as ChakraCardFooter,
  CardFooterProps as ChakraCardFooterProps,
} from "@chakra-ui/card";
import { BaseMultiStyleConfiguration } from "../../../theme";
import { Avatar, Spinner, Text } from "../../Atoms";
import { useHover } from "../../../hooks";

export const cardPartsList: (
  | "body"
  | "header"
  | "footer"
  | "container"
  | "avatar"
  | "title"
  | "description"
  | "content"
)[] = [
  "body",
  "header",
  "footer",
  "container",
  "avatar",
  "title",
  "description",
  "content",
];

type Parts = typeof cardPartsList;

export interface CardThemeConfig extends BaseMultiStyleConfiguration<Parts> {}

export interface CardProps extends ChakraCardProps {
  headerProps?: ChakraCardHeaderProps;
  bodyProps?: ChakraCardBodyProps;
  footerProps?: ChakraCardFooterProps;
  header?: React.ReactNode;
  footer?: React.ReactNode;
  isLoading?: boolean;
  img?: string;
  title?: string;
  description?: string;
}

export const Card = ({
  headerProps,
  bodyProps,
  footerProps,
  header,
  footer,
  isLoading,
  img,
  ...props
}: CardProps) => {
  const [bodyRef, isHoveredBody] = useHover();
  const [avatarRef, isHoveredAvatar] = useHover();
  const styles = useChakraMultiStyleConfig("Card", {
    isLoading,
    isHovered: isHoveredBody || isHoveredAvatar,
  });

  return (
    <ChakraCard {...props} sx={styles.container}>
      <ChakraCardHeader sx={styles.header} {...headerProps}>
        <Avatar
          ref={avatarRef}
          src={img}
          isLoading={isLoading}
          size='lg'
          sx={styles.avatar}
        />
      </ChakraCardHeader>

      <ChakraCardBody ref={bodyRef} sx={styles.body} {...bodyProps}>
        {isLoading ? <Spinner /> : <Content {...props} />}
      </ChakraCardBody>
      {footer && !isLoading && (
        <ChakraCardFooter sx={styles.footer} {...footerProps}>
          {footer}
        </ChakraCardFooter>
      )}
    </ChakraCard>
  );
};

export interface ContentProps
  extends Pick<CardProps, "title" | "description"> {}

export const Content = ({ title, description }: ContentProps) => {
  const styles = useChakraMultiStyleConfig("Card");
  return (
    <ChakraVStack sx={styles.content}>
      <Text sx={styles.title}>{title}</Text>
      <Text sx={styles.description}>{description}</Text>
    </ChakraVStack>
  );
};
