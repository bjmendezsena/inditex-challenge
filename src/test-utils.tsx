import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
