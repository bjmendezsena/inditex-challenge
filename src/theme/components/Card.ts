import { CardThemeConfig, cardPartsList } from "../../components";

export const ConfigCard: CardThemeConfig = {
  parts: cardPartsList,
  baseStyle: ({ isHovered }) => ({
    container: {
      bg: "transparent",
      borderRadius: "md",
      boxShadow: "none",
      p: 4,
      maxW: "230px",
    },
    body: {
      overflowY:'hidden',
      transition: "all 0.02s ease-in-out",
      mb: 4,
      bg: "white",
      borderRadius: "md",
      boxShadow: "md",
      p: 4,
      minW: "150px",
      minH: "120px",
      display: "flex",
      justifyContent: "center",
      pt: 12,
      ...(isHovered && {
        cursor: "pointer",
        bg: "gray.50",
      }),
    },
    header: {
      position: "relative",
      fontSize: "xl",
    },
    footer: {
      fontSize: "sm",
      display: "flex",
      justifyContent: "center",
    },
    avatar: {
      bg: "gray.100",
      _hover: {
        cursor: "pointer",
      },
    },
    title: {
      fontSize: "15px",
      fontWeight: "semibold",
      color: "black",
    },
    description: {
      fontSize: "12px",
      fontWeight: "regular",
      color: "text.bold",
      textOverflow:'ellipsis',
    },
  }),
};
