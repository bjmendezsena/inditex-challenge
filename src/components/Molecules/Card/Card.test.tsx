import { screen } from "@testing-library/react";

import { render } from "../../../test-utils";
import { Card, CardProps } from "./Card";
import userEvent from "@testing-library/user-event";

const src = "https://placekitten.com/200/200";

const onClick = jest.fn();
const defaultProps: CardProps = {
  title: "Card title",
  description: "Card description",
  onClick: onClick,
};

describe(`<${Card.name}/> component`, () => {
  const factoryComponent = (props?: CardProps) =>
    render(<Card {...defaultProps} {...props} data-testid={Card.name} />);
  test("renders correctly", () => {
    const view = factoryComponent();
    expect(view.asFragment()).toMatchSnapshot();
  });
  test("renders correctly with header", () => {
    const view = factoryComponent({ header: "header" });
    expect(view.asFragment()).toMatchSnapshot("WithHeader");
  });
  test("renders correctly with isLoading", () => {
    const view = factoryComponent({ isLoading: true });
    expect(view.asFragment()).toMatchSnapshot("WithIsLoading");
  });
  test("renders correctly with img", () => {
    const view = factoryComponent({ img: src });
    expect(view.asFragment()).toMatchSnapshot("WithImg");
  });
  test("should call onClick function", () => {
    factoryComponent();
    userEvent.click(screen.getByTestId(Card.name));
    expect(onClick).toHaveBeenCalled();
  });
});
