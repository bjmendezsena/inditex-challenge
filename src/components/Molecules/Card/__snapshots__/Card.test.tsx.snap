// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-card css-1jvhf43"
    data-testid="Card"
    description="Card description"
    title="Card title"
  >
    <div
      class="chakra-card__header css-18f68ux"
    >
      <div
        class="css-14w3rw9"
      >
        <span
          class="chakra-avatar css-gi3kny"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="chakra-card__body css-bwepl0"
    >
      <div
        class="chakra-stack css-tl3ftk"
      >
        <p
          class="css-1qbkn0c"
        >
          Card title
        </p>
        <p
          class="css-81w022"
        >
          Card description
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Card/> component renders correctly with header: WithHeader 1`] = `
<DocumentFragment>
  <div
    class="chakra-card css-1jvhf43"
    data-testid="Card"
    description="Card description"
    title="Card title"
  >
    <div
      class="chakra-card__header css-18f68ux"
    >
      <div
        class="css-14w3rw9"
      >
        <span
          class="chakra-avatar css-gi3kny"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="chakra-card__body css-bwepl0"
    >
      <div
        class="chakra-stack css-tl3ftk"
      >
        <p
          class="css-1qbkn0c"
        >
          Card title
        </p>
        <p
          class="css-81w022"
        >
          Card description
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Card/> component renders correctly with img: WithImg 1`] = `
<DocumentFragment>
  <div
    class="chakra-card css-1jvhf43"
    data-testid="Card"
    description="Card description"
    title="Card title"
  >
    <div
      class="chakra-card__header css-18f68ux"
    >
      <div
        class="css-14w3rw9"
      >
        <span
          class="chakra-avatar css-gi3kny"
        >
          <svg
            aria-label=" avatar"
            class="chakra-avatar__svg css-16ite8i"
            role="img"
            viewBox="0 0 128 128"
          >
            <path
              d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
              fill="currentColor"
            />
            <path
              d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="chakra-card__body css-bwepl0"
    >
      <div
        class="chakra-stack css-tl3ftk"
      >
        <p
          class="css-1qbkn0c"
        >
          Card title
        </p>
        <p
          class="css-81w022"
        >
          Card description
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Card/> component renders correctly with isLoading: WithIsLoading 1`] = `
<DocumentFragment>
  <div
    class="chakra-card css-1jvhf43"
    data-testid="Card"
    description="Card description"
    title="Card title"
  >
    <div
      class="chakra-card__header css-18f68ux"
    >
      <div
        class="css-14w3rw9"
      >
        <div
          class="chakra-skeleton css-1536nqa"
        />
      </div>
    </div>
    <div
      class="chakra-card__body css-bwepl0"
    >
      <div
        class="chakra-spinner css-18f4w0n"
        data-testid="spinner"
      >
        <span
          class="css-8b45rq"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
